### PR TITLE
[DO NOT REVIEW]

### PR DESCRIPTION
**Pull request by linzebing**
[DO NOT REVIEW]

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/SystemSessionProperties.java [hunk 1]
```diff
@@ -813,10 +813,10 @@ public SystemSessionProperties(
                         "Target size in bytes of all task inputs for a single fault tolerant task",
                         queryManagerConfig.getFaultTolerantExecutionTargetTaskInputSize(),
                         false),
-                integerProperty(
-                        FAULT_TOLERANT_EXECUTION_TARGET_TASK_SPLIT_COUNT,
-                        "Target number of splits for a single fault tolerant task (split weight aware)",
-                        queryManagerConfig.getFaultTolerantExecutionTargetTaskSplitCount(),
+                dataSizeProperty(
+                        FAULT_TOLERANT_EXECUTION_STANDARD_SPLIT_SIZE,
+                        "Standard split size for a single fault tolerant task (split weight aware)",
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/SystemSessionProperties.java:848]: nit: define description as constant in Config class and use it here

File: core/trino-main/src/main/java/io/trino/SystemSessionProperties.java [hunk 2]
```diff
@@ -162,7 +162,12 @@
     public static final String RETRY_DELAY_SCALE_FACTOR = "retry_delay_scale_factor";
     public static final String LEGACY_MATERIALIZED_VIEW_GRACE_PERIOD = "legacy_materialized_view_grace_period";
     public static final String HIDE_INACCESSIBLE_COLUMNS = "hide_inaccessible_columns";
-    public static final String FAULT_TOLERANT_EXECUTION_TARGET_TASK_INPUT_SIZE = "fault_tolerant_execution_target_task_input_size";
+    public static final String FAULT_TOLERANT_EXECUTION_ADAPTIVE_TASK_SIZING_STEP_SIZE = "fault_tolerant_execution_adaptive_task_sizing_step_size";
+    public static final String FAULT_TOLERANT_EXECUTION_ADAPTIVE_TASK_SIZING_GROWTH_FACTOR = "fault_tolerant_execution_adaptive_task_sizing_growth_factor";
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/SystemSessionProperties.java:166]: Hmm - I would have those separate for writer and arbitrary too. For arbitrary we probalby want 2x less frequently. And for writers some smaller factor but more often for more smooth growth.

File: core/trino-main/src/main/java/io/trino/SystemSessionProperties.java [hunk 3]
```diff
@@ -808,10 +813,35 @@ public SystemSessionProperties(
                         featuresConfig.isHideInaccessibleColumns(),
                         value -> validateHideInaccessibleColumns(value, featuresConfig.isHideInaccessibleColumns()),
                         false),
+                integerProperty(
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/SystemSessionProperties.java:816]: those are usefult for experimenting - but maybe we should declare them as hidden?

File: core/trino-main/src/main/java/io/trino/SystemSessionProperties.java [hunk 4]
```diff
@@ -162,7 +162,12 @@
     public static final String RETRY_DELAY_SCALE_FACTOR = "retry_delay_scale_factor";
     public static final String LEGACY_MATERIALIZED_VIEW_GRACE_PERIOD = "legacy_materialized_view_grace_period";
     public static final String HIDE_INACCESSIBLE_COLUMNS = "hide_inaccessible_columns";
-    public static final String FAULT_TOLERANT_EXECUTION_TARGET_TASK_INPUT_SIZE = "fault_tolerant_execution_target_task_input_size";
+    public static final String FAULT_TOLERANT_EXECUTION_ADAPTIVE_TASK_SIZING_STEP_SIZE = "fault_tolerant_execution_adaptive_task_sizing_step_size";
+    public static final String FAULT_TOLERANT_EXECUTION_ADAPTIVE_TASK_SIZING_GROWTH_FACTOR = "fault_tolerant_execution_adaptive_task_sizing_growth_factor";
+    public static final String FAULT_TOLERANT_EXECUTION_NON_WRITER_TASK_INIT_TARGET_INPUT_SIZE = "fault_tolerant_execution_non_writer_task_init_target_input_size";
+    public static final String FAULT_TOLERANT_EXECUTION_NON_WRITER_TASK_MAX_INPUT_SIZE = "fault_tolerant_execution_non_writer_task_max_input_size";
+    public static final String FAULT_TOLERANT_EXECUTION_WRITER_TASK_INIT_TARGET_INPUT_SIZE = "fault_tolerant_execution_writer_task_init_target__input_size";
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/SystemSessionProperties.java:169]: nit: typo double underscore

File: core/trino-main/src/main/java/io/trino/SystemSessionProperties.java [hunk 5]
```diff
@@ -162,7 +162,12 @@
     public static final String RETRY_DELAY_SCALE_FACTOR = "retry_delay_scale_factor";
     public static final String LEGACY_MATERIALIZED_VIEW_GRACE_PERIOD = "legacy_materialized_view_grace_period";
     public static final String HIDE_INACCESSIBLE_COLUMNS = "hide_inaccessible_columns";
-    public static final String FAULT_TOLERANT_EXECUTION_TARGET_TASK_INPUT_SIZE = "fault_tolerant_execution_target_task_input_size";
+    public static final String FAULT_TOLERANT_EXECUTION_ADAPTIVE_TASK_SIZING_STEP_SIZE = "fault_tolerant_execution_adaptive_task_sizing_step_size";
+    public static final String FAULT_TOLERANT_EXECUTION_ADAPTIVE_TASK_SIZING_GROWTH_FACTOR = "fault_tolerant_execution_adaptive_task_sizing_growth_factor";
+    public static final String FAULT_TOLERANT_EXECUTION_NON_WRITER_TASK_INIT_TARGET_INPUT_SIZE = "fault_tolerant_execution_non_writer_task_init_target_input_size";
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/SystemSessionProperties.java:167]: init -> initial

File: core/trino-main/src/main/java/io/trino/SystemSessionProperties.java [hunk 6]
```diff
@@ -162,7 +162,12 @@
...
+    public static final String FAULT_TOLERANT_EXECUTION_NON_WRITER_TASK_MAX_INPUT_SIZE = "fault_tolerant_execution_non_writer_task_max_input_size";
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/SystemSessionProperties.java:168]: max_target

File: core/trino-main/src/main/java/io/trino/SystemSessionProperties.java [hunk 7]
```diff
@@ -162,7 +162,12 @@
...
+    public static final String FAULT_TOLERANT_EXECUTION_WRITER_TASK_INIT_TARGET_INPUT_SIZE = "fault_tolerant_execution_writer_task_init_target__input_size";
+    public static final String FAULT_TOLERANT_EXECUTION_WRITER_TASK_MAX_INPUT_SIZE = "fault_tolerant_execution_writer_task_max_input_size";
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/SystemSessionProperties.java:170]: max_target

File: core/trino-main/src/main/java/io/trino/execution/QueryManagerConfig.java [hunk 8]
```diff
@@ -625,20 +630,88 @@ public QueryManagerConfig setRemoteTaskGuaranteedSplitPerTask(int remoteTaskGuar
...
+    @ConfigDescription("Initial target input size for non-writer fault tolerant tasks")
+    public QueryManagerConfig setFaultTolerantExecutionNonWriterTaskInitTargetInputSize(DataSize faultTolerantExecutionNonWriterTaskInitTargetInputSize)
+    {
+        this.faultTolerantExecutionNonWriterTaskInitTargetInputSize = faultTolerantExecutionNonWriterTaskInitTargetInputSize;
+        return this;
+    }
+
     @NotNull
-    public DataSize getFaultTolerantExecutionTargetTaskInputSize()
+    public DataSize getFaultTolerantExecutionNonWriterTaskMaxInputSize()
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/execution/QueryManagerConfig.java:674]: MaxTarget

File: core/trino-main/src/main/java/io/trino/execution/QueryManagerConfig.java [hunk 9]
```diff
@@ -625,20 +630,88 @@ public QueryManagerConfig setRemoteTaskGuaranteedSplitPerTask(int remoteTaskGuar
...
+    {
+        return faultTolerantExecutionNonWriterTaskMaxInputSize;
+    }
+
+    @Config("fault-tolerant-execution-non-writer-task-max-input-size")
+    @ConfigDescription("Maximum input size for non-writer fault tolerant tasks")
+    public QueryManagerConfig setFaultTolerantExecutionNonWriterTaskMaxInputSize(DataSize faultTolerantExecutionNonWriterTaskMaxInputSize)
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/execution/QueryManagerConfig.java:681]: MaxTarget

Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/execution/QueryManagerConfig.java:681]: and elsewhere :)

File: core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenTaskSourceFactory.java [hunk 10]
```diff
@@ -189,7 +213,8 @@ private SplitAssigner createSplitAssigner(
                     sourcePartitioningScheme,
                     outputDataSizeEstimates,
                     fragment,
-                    getFaultTolerantExecutionTargetTaskInputSize(session).toBytes());
+                    // number of tasks are bounded for hash distribution so adaptive task sizing is automatic
+                    partitioning.equals(SCALED_WRITER_HASH_DISTRIBUTION) ? writerTaskInitTargetInputSizeInBytes : nonWriterTaskInitTargetInputSizeInBytes);
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenTaskSourceFactory.java:217]: We should have adaptive sizing for `HashDistributionSplitAssigner` where ware dealing with writer fragment.
In that case we splitting single hash partitions which can be huge into multiple tasks. And logic should be the same. Initially crated tasks should get less data and task created later should get more data.

I guess you can parametrize `HashDistributionSplitAssigner` with same set of parameters as `ArbitraryDistributionSplitAssigner` but make the "growth" ones optional. That is
```
            OptionalInt adaptiveTaskSizingStepSize,
            OptionalDouble adaptiveTaskSizingGrowthFactor,
            long initTargetPartitionSizeInBytes,
            OptionalLong maxPartitionSizeInBytes,
```

Then the Optional parameters will be set and used only for writer case.

Or maybe you can skip Optionals and just use initial and max same for non-writer case.

File: core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenTaskSourceFactory.java [hunk 11]
```diff
@@ -170,13 +179,28 @@ private SplitAssigner createSplitAssigner(
...
+                    partitionedSources,
+                    replicatedSources,
+                    standardSplitSizeInBytes,
+                    adaptiveTaskSizingStepSize,
+                    adaptiveTaskSizingGrowthFactor,
+                    nonWriterTaskInitTargetInputSizeInBytes,
+                    nonWriterTaskMaxInputSizeInBytes,
+                    maxArbitraryDistributionTaskSplitCount);
+        }
+        if (partitioning.equals(SCALED_WRITER_ROUND_ROBIN_DISTRIBUTION)) {
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenTaskSourceFactory.java:194]: IDK when this one is used. @arhimondr ?

File: core/trino-main/src/main/java/io/trino/SystemSessionProperties.java (hunk 1 above)
Comment (linzebing) [reply core/trino-main/src/main/java/io/trino/SystemSessionProperties.java:848]: You mean `QueryManagerConfig.java`?

**Pull request closed without merging**